File: C++/Tree/Binary Tree Level Order Traversal.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

/**
 * https://leetcode.com/problems/binary-tree-level-order-traversal/
 * */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode *root) {
        if (!root) return {};
        vector<vector<int>> ans;
        queue<TreeNode *> queue;
        queue.push(root);
        while (!queue.empty()) {
            vector<int> level;
            for (int i = queue.size() - 1; i >= 0; i--) {
                auto front = queue.front();
                if (front) {
                    level.push_back(front->val);
                    queue.push(front->left);
                    queue.push(front->right);
                }
                queue.pop();
            }
            if (!level.empty())
                ans.push_back(level);
        }
        return ans;
    }
};

File: Java-Kotlin/src/main/kotlin/tree/Binary Tree Level Order Traversal.kt
package tree

import java.util.*
import kotlin.collections.ArrayList

fun levelOrder(root: TreeNode?): List<List<Int>> {
    val ans = ArrayList<ArrayList<Int>>()
    val queue: Queue<TreeNode> = LinkedList()
    queue.add(root)
    while (!queue.isEmpty()) {
        val row = ArrayList<Int>()
        for (i in queue.size downTo 1) {
            val front = queue.poll()
            if (front != null) {
                row.add(front.`val`)
                queue.add(front.left)
                queue.add(front.right)
            }
        }
        if (row.size > 0) ans.add(row)
    }
    return ans
}
